### ISSUE TITLE
NCCL Timeout if Number of Batches are not same per rank for Eval mode

### ISSUE THREAD

**Issue by BackNode** [meta-pytorch/torchrec#1014]
Hi all,

I am following [this example](https://github.com/facebookresearch/recipes/tree/main/torchrecipes/rec) to run TorchRec with pytorch_lightning. However it will hangs till NCCL timeout if number of batches are different during **validation** stage. 

I know #Batches should be same for training mode, while it does not make sense for eval mode. Did I miss something? Appreciate for your help.

Code Snippet:
```Python
class ShardedModelWrapper(LightningModule):
    """
    A wrapper for a distributed model parallel model (Beta). This is still under active development.
    """

    def __init__(
        self,
        model: nn.Module,
        loss: torch.nn.Module,
    ):
        super().__init__()

        ...

        self.dist_model = DistributedModelParallel(
            module=model,
            sharders=cast(List[ModuleSharder[nn.Module]], sharders),
            plan=plan,
            env=env,
            device=device,
        )

        # TODO: train_pipeline, val_pipeline, test_pipeline
        self.pipelines = {}
        train_pipeline: TrainPipelineSparseDist = TrainPipelineSparseDist(
            model=self.dist_model,
            optimizer=self.get_dense_optimizer(),
            device=device,
        )
        self.pipelines["train"] = train_pipeline
        val_pipeline: TrainPipelineSparseDist = TrainPipelineSparseDist(
            model=self.dist_model,
            optimizer=None,
            device=device,
        )
        self.pipelines["val"] = val_pipeline

        # This is important!
        # skip optimization in pl, since it's done in TorchRec train_pipeline
        self.automatic_optimization = False

    def _step(
        self,
        dataloader_iter: Any,
        batch_idx: int,
        phase_type: str,
    ) -> torch.Tensor:
        loss, logits, labels, ids, rguids = self.pipelines[phase_type].progress(dataloader_iter)
        return {"loss": loss}

    def training_step(
        self,
        dataloader_iter: Iterator[In],
        batch_idx: int,
        *args: Any,
        **kwargs: Any,
    ) -> None:
        loss_dict = self._step(dataloader_iter, batch_idx, "train")

        sch = self.lr_schedulers()
        sch.step()
        return loss_dict

    def validation_step(
        self,
        dataloader_iter: Iterator[In],
        batch_idx: int,
        *args: Any,
        **kwargs: Any,
    ) -> Dict[str, torch.Tensor]:
        return self._step(dataloader_iter, batch_idx, "val")

    def on_train_epoch_start(self) -> None:
        self.pipelines["train"]._model.train()

    def on_validation_epoch_start(self) -> None:
        self.pipelines["val"]._model.eval()
```

**Comment by BackNode** [comment]
@colin2328 Have you met this? Any clue?

**Comment by bigning** [comment]
What do you mean by "Number of Batches are not same per rank"? Do you mean rank_0 run m batches while rank_1 run n batches? With pytorch DMP on multiple ranks, it needs collective communication across ranks for each iteration for both forward and backward. So if on some ranks it's not running the n-th batch while the rest of the ranking are running the n-th batch, the collective communication will hang.

**Comment by BackNode** [comment]
>Do you mean rank_0 run m batches while rank_1 run n batches?

Yes, that's what I mean.

Even for validation or predicting mode, we need to ensure same number of batches on all ranks? It's hard to guarantee that when we are not aware of accurate data size.
